Rework `analyze` function signature

vrugtehagel opened the pull request:
The original `analyze` function takes a `visitor` argument, and this is a bit awkward to work with in isolation. To do performance optimizations, it is more ergonomic for this function to be lower level, taking only primitive arguments, reducing the entanglement with other parts of the codebase.

As such, I've reworked it to a `topLevel` function that reads a string of JavaScript until it encounters a character that "breaks" out of the top-level context; specifically this happens when it encounters a top-level `}`, `]`, or `|>`. The algorithm remains mostly the same, since conceptually the `analyze` function and `topLevel` function are doing very similar things.

The "include" tag here is a bit of an outlier. Its syntax is actually rather complex to get right, e.g. `include resolve({path: 'foo.tmpl'}) { bar: 'baz' })` is difficult to parse (both with `analyze` as well as `topLevel`). In this rework I've assumed that the object passed as data is passed literally, i.e. starts with `{` and ends with a matching `}`. I think that's a reasonable requirement to avoid the ambiguity. I've added a test here that passes on this branch but would fail on `main`.

vrugtehagel commented:
Oh, cool! That's an interesting (and creative) idea.

The reason I went with this signature is that it'll be easier to do optimizations with. It is one of the lowest level functions in the codebase - after meriyah, it probably is the heaviest function. I feel like working with primitives is probably what JS engines are most optimized for, so I'm trying to keep the signature simple for that reason. It'll be much easier to change the logic of this function like this, which is what I'm guessing will be the most impactful for Vento's compilation speed.

If we were to use a generator, we probably `yield` more than we need to - and it also prevents some potential optimizations that would skip certain tokens. For example, imagine a shortcut skip for a simple array literal; we then have to manually handle yielding the brackets and potentially tokens, slowing things back down, potentially even preventing us from making such optimizations.

I do get the argument that it makes usage a bit more verbose, but ultimately it is by no means user-facing and I think we want be focusing on getting it to be really fast rather than nice to work with. This is also why I don't think parsing `(` and `)` is preferable; it would make parsing slower for very little benefit.

The `include` logic I went for here is doing _some_ trial-and-error but only does unnecessary work if an object is provided _and_ the path is written using a template literal with interpolation. And this is realistically unavoidable, even if we have handling for `()` because users could write e.g. `{{ include {foo: "bar.tmpl"}.foo { name: 'name' } }}`. The syntax is simply a bit awkward to parse, but I think this method is fine in terms of performance given the constraints.

What are your thoughts on this? Perhaps it's also a balance to strike between maintainability of Vento's source and performance, and I guess it's somewhat subjective where on that scale you'd want Vento to end up.

oscarotero commented:
> If we were to use a generator, we probably yield more than we need to - and it also prevents some potential optimizations that would skip certain tokens.

Yield approach would work in a similar way as topLevel(). It only returns the top level stuff, not every open/close brackets. For example `{ {foo: {bar}} }` would yield only the last closing bracket (in the same way as the current analyzer). So this would allows to create shortcuts to skip some tokens.

I'd like to have a balance between performance and code clarity, specially if the analyzer function is not the bottleneck at this moment. Premature optimization is never a good idea and to me this code is confusing:

```js
if (tagCode.endsWith("}")) {
  const target = tagCode.length - 1;
  let index = -1;
  for (const match of tagCode.matchAll(/{/g)) {
    if (topLevel(tagCode, match.index + 1) == target) {
      index = match.index;
      break;
    }
  }
```

It looks like it checks whether the code ends with `}` and then searches by all `{` in the code to find the first occurrence paired with the latest character? 

For example, I'm not sure if it would work with this, because it doesn't detect if the character is inside a quote, a comment, etc:
```vto
{{ "file{name.vto" /* this is { a comment */ { data } }}
```
Or even if it works, it would be slower because it has to parse the code for every character until it find the correct one.
I think it's more robust to parse the code only once, taking advance of the analyzer logic to prevent false positives, instead of searching for raw characters in the code. The `topLevel` function seems to return only when a `}` or `]` is found in the root level or when it reaches to the end of the code, which is something we don't need. We need to detect when a new bracket is open or close at the root level, to split the code in different parts.

> I don't think parsing ( and ) is preferable; it would make parsing slower for very little benefit.

It would make the parsing more accurate. I think the impact in the performace would be minimum, but we can measure. It's easy to implement.


I'm a bit surprised about the benchmark. Are you sure both versions do the same thing? I can expect some improvement but 90x faster?

vrugtehagel commented:
I've reworked it to be a generator, and added `yield` statements when it encounters a `{` or `[`. The extra yields are useful for the "include" tag parsing (the last top-level `{` is the data object) but are very inelegant with the non-generator version of `topLevel()`. I originally opened a separate PR (https://github.com/ventojs/vento/pull/122) to compare the two methods, but decided against it because the non-generator function got too messy.

vrugtehagel commented on src/js.ts:
```diff
@@ -130,16 +126,15 @@ export default function analyze(source: string, visitor: Visitor) {
       }
 
       case "]": {
+        if (statuses.length === 0) yield [index - 1, "]"];
```
You know what? You're so right. I actually wrote it like this intentionally, but I guess I was a bit stuck in this idea as I applied the same concept to Sublime's syntax definition. But you're absolutely correct that it doesn't make sense here. Thanks for the sharp eye!

Should be better now; it yields on any top-level open/close bracket (including "higher" brackets, like in `foo]`). Also adjusted the relevant parts in the plugins and tokenizer and it does simplify things. You love to see it!

oscarotero commented on src/js.ts:
```diff
@@ -42,9 +43,7 @@ export default function analyze(source: string, visitor: Visitor) {
-            if (!statuses.length && visitor("open-bracket", index) === false) {
-              return;
-            }
+            if (statuses.length == 0) yield [index - 1, "{"];
```
For opening brackets, it returns the index before, and for closing brackets and filters, the index after.
I understand the reason, but I'm not sure if this can be confusing. The previous `analyze` function returns always the position after the character was found (that maybe it wasn't correct but at least consistent).

It's not a strong opinion, but I think that maybe expected result is the position before the character starts, to mimic the behavior of indexOf. For example:

```js
//for [bar] of { foo }

[4, "["]
[8, "]"]
[13, "{"]
[19, "}"]
[20, ""]
```
What do you think?

vrugtehagel replied:
Yeah, I also thought about this - it does feel somewhat inconsistent, but in the end I went with what I felt was easiest to use. When you're parsing a piece of JS and you're looking for opening brackets, you probably want to cut _before_ the bracket. If you are looking for a closing bracket, probably you want to cut the string _after_ the bracket. For `|>` it is a bit ambiguous, because it's more of a separator, but here too I went with ease-of-use and return the index after.

I don't have a strong opinion about it either, though. I'm fine going for consistency and always return the index after the "cause" of yielding, if you prefer. The tradeoff is having to do `index - 1` for matching most opening brackets 🤷

oscarotero replied:
I think it should return the position **before** the caracter. This is the behavior of other native functions to search in strings, like `indexOf`. For example:

```js
const code = "Hello { world }";

const position1 = code.indexOf("{"); // 6

for (const [position, cause] of iterateTopLevel(code)) {
  console.log(position, cause); // [6, "{"]
}
```

Or to be more clear:

```js
for (const [position, cause] of iterateTopLevel(code)) {
  let cause2 = code.slice(position, cause.length);
  assert(cause === cause2);
}
```

oscarotero commented on src/js.ts:
```diff
@@ -15,12 +15,13 @@ type status =
-type Visitor = (type: breakpoints, index: number) => false | void;
-
-export default function analyze(source: string, visitor: Visitor) {
+export function* topLevel(
```
I'm not sure about the named export. `analyze` is not a good name, but `topLevel` is not descriptive enough. And due it's a module with a single function, probably should be exported as `default`.

vrugtehagel replied:
Oh, I can change it back to a default export - just usually go for named ones myself and didn't catch it here.

I agree about the name - what would you suggest here? Perhaps `iterateTopLevel()`?

oscarotero reviewed (commented):
I love the new approach! Thanks so much for the work!
I just left a couple of questions.

vrugtehagel commented on src/tokenizer.ts:
```diff
@@ -102,20 +102,14 @@ export default function tokenize(source: string): TokenizeResult {
-
-  return indexes;
+      continue;
+    } else if (!source.startsWith("}}", index - 1)) continue;
```
It's not always `false` - not the second argument to `startsWith()` here. It denotes the index to start the matched string at. It doesn't read particularly well (as is evident here) but it is the appropriate choice for this check. As for just looking for `}`, it depends a bit on what you prefer to happen when parsing e.g. `{{ foo } bar }}` - if we stop looking after the first `}` then I suppose `foo` gets printed and then a literal `" bar }}"`, probably not what is intended. So we need to make sure the stopping point of the tag really does match `}}`, hence this `startsWith()` here.

oscarotero replied:
Ok, I understand now.

In the original `analyze` function `{{ foo } bar }}` returned `{{ foo } bar }` because it stops when the two opening brackets are closed, but it doesn't check whether these brackets are consecutive. Your approach seems to be more accurate.
